# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/jsapi/PeerConnectionImpl.cpp
# Commit: a55a718715f6
# Full Hash: a55a718715f69cba1ff9fbae4787881b4d930907
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 3
# Description:
#   Bug 1654112 - Report video frame-/bitrate telemetry once per call instead of every second. r=ng
#   
#   This is similar to the behavior prior to bug 1532898, which introduced telemetry
#   reporting every second. There's a bit of a difference in that we now only report
#   data if we had collected some through the 1 second interval telemetry timer in
# ==============================================================================

diff -r 069c5baf6d32 -r a55a718715f6 dom/media/webrtc/jsapi/PeerConnectionImpl.cpp
--- a/dom/media/webrtc/jsapi/PeerConnectionImpl.cpp	Thu Jan 28 10:54:32 2021 +0100
+++ b/dom/media/webrtc/jsapi/PeerConnectionImpl.cpp	Mon Jan 25 16:54:14 2021 +0100
@@ -2851,28 +2851,24 @@
       });
 }
 
-void PeerConnectionImpl::RecordConduitTelemetry() {
+void PeerConnectionImpl::CollectConduitTelemetryData() {
+  MOZ_ASSERT(NS_IsMainThread());
+
   if (!mMedia) {
     return;
   }
 
   nsTArray<RefPtr<VideoSessionConduit>> conduits;
   for (const auto& transceiver : mMedia->GetTransceivers()) {
-    RefPtr<MediaSessionConduit> conduit = transceiver->GetConduit();
-    if (conduit) {
-      auto asVideo = conduit->AsVideoSessionConduit();
-      if (asVideo) {
-        conduits.AppendElement(asVideo.value());
-      }
+    if (RefPtr<MediaSessionConduit> conduit = transceiver->GetConduit()) {
+      conduit->AsVideoSessionConduit().apply(
+          [&](const auto& aVideo) { conduits.AppendElement(aVideo); });
     }
   }
 
-  mSTSThread->Dispatch(
-      NS_NewRunnableFunction(__func__, [conduits = std::move(conduits)]() {
-        for (const auto& conduit : conduits) {
-          conduit->RecordTelemetry();
-        }
-      }));
+  for (const auto& conduit : conduits) {
+    conduit->CollectTelemetryData();
+  }
 }
 
 template <class T>